File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.java
class Solution {
    public int solution(int[] bandage, int health, int[][] attacks) {
        int curHealth = health;

        for (int i = 0; i < attacks.length; i++) {

            if (i > 0) {
                //bandage 시간을 재서 체력 회복
                int bandageTime = attacks[i][0] - attacks[i - 1][0] - 1;
                curHealth += bandageTime * bandage[1];

                //추가 체력 회복 시간을 재서 추가 회복
                if (bandageTime >= bandage[0]) {
                    curHealth += (bandageTime / bandage[0]) * bandage[2];
                }

                //체력이 시작 체력보다 많으면 변경
                if (curHealth > health)
                    curHealth = health;
            }

            //공격 받음
            curHealth -= attacks[i][1];
            if (curHealth <= 0) 
                return -1;
        }
        return curHealth;
    }
}

File: 프로그래머스/1/250137. ［PCCP 기출문제］ 1번 ／ 붕대 감기/［PCCP 기출문제］ 1번 ／ 붕대 감기.js
function solution(bandage, health, attacks) {
    const maxHealth = health;
    let extraHealth = health;
    let turn = 0;
    let count = -1;
    while(true){
        if(attacks.length < 1 ){
            break;
        }
        let isAttacked = false;
        let attackedData = null;
        if(attacks[0][0] === turn){
            isAttacked = true;
            count = 0;
            attackedData = attacks.shift();
        }
        if(isAttacked){
            extraHealth = extraHealth - attackedData[1];
            if(extraHealth < 1 ){
                break;
            }
        }else{
            count = count + 1;
            if(maxHealth > extraHealth){
                if(count === bandage[0]){
                    extraHealth = extraHealth + bandage[1] + bandage[2]
                    count = 0;
                }else {
                    extraHealth = extraHealth + bandage[1]     
                }
            }
            if(maxHealth <= extraHealth) extraHealth = health
        }

        console.log(turn, extraHealth, count, isAttacked)
        turn = turn + 1;
    }


    return extraHealth > 0 ? extraHealth : -1;
}